src/ORM/Query.php

```diff
@@ -0,0 +1,240 @@
+<?php
+
+namespace Lampager\Cake\ORM;
+
+use Cake\Database\Expression\OrderByExpression;
+use Cake\Database\Expression\OrderClauseExpression;
+use Cake\Database\Expression\QueryExpression;
+use Cake\Database\ExpressionInterface;
+use Cake\ORM\Query as BaseQuery;
+use Lampager\Cake\Paginator;
+use Lampager\Contracts\Cursor;
+
+/**
+ * @method $this forward(bool $forward = true) Define that the current pagination is going forward.
+ * @method $this backward(bool $backward = true) Define that the current pagination is going backward.
+ * @method $this exclusive(bool $exclusive = true) Define that the cursor value is not included in the previous/next result.
+ * @method $this inclusive(bool $inclusive = true) Define that the cursor value is included in the previous/next result.
+ * @method $this seekable(bool $seekable = true) Define that the query can detect both "has_previous" and "has_next".
+ * @method $this unseekable(bool $unseekable = true) Define that the query can detect only either "has_previous" or "has_next".
+ * @method $this fromArray(mixed[] $options) Define options from an associative array.
+ */
+class Query extends BaseQuery
+{
+    /** @var Paginator */
+    protected $_paginator;
+
+    /** @var Cursor|int[]|string[] */
+    protected $_cursor = [];
+
+    /**
+     * Construct query.
+     *
+     * @param \Cake\Database\Connection $connection The connection object
+     * @param \Cake\ORM\Table           $table      The table this query is starting on
+     */
+    public function __construct($connection, $table)
+    {
+        parent::__construct($connection, $table);
+
+        $this->_paginator = Paginator::create($this);
+    }
+
+    /**
+     * Create query based on the existing query. This factory copies the internal
+     * state of the given query to a new instance.
+     *
+     * @param BaseQuery $query
+     */
+    public static function fromQuery($query)
+    {
+        $obj = new static($query->getConnection(), $query->getRepository());
+
+        foreach (get_object_vars($query) as $k => $v) {
```

Comment:
How about `clone` operator? It just behaves shallow copying.

Reply:
This method copies all the properties from the given `\Cake\ORM\Query` to a `\Lampager\Cake\ORM\Query` that extends it. How can it be written with `clone` operator while they differ in their types?

Reply:
That was my misunderstanding, okay